text/0000-test-framework-binnacle.md

```diff
@@ -0,0 +1,357 @@
+
+LOOP_DATA="""
+[
+  {
+    "node": "$N1",
+    "device": "/dev/vdc"
+  },
+  {
+    "node": "$N2",
+    "device": "/dev/vdc"
+  },
+  {
+    "node": "$N3",
+    "device": "/dev/vdc"
+  }
+]
+"""
+LOOP:
+    NODE=$node
+    TEST kube_node_setup
+    TEST join_kube_master tmp/master-details.json
+    TEST loop_device_setup $device
+
+NODE=$MASTER
+TEST kubectl create -f $OPERATOR_YAML
+EXPECT_WITHIN -t 180 -v 3 kubectl get pods -n kadalu | grep "Running" | wc -l
+TEST kubectl kadalu storage-add $STORAGE_POOL_NAME --type Replica3 \
+    --device $N1:/dev/vdc
+    --device $N2:/dev/vdc
+    --device $N3:/dev/vdc
+EXPECT_WITHIN -t 180 -v 3 kubectl get pods -n kadalu | grep $STORAGE_POOL_NAME | wc -l
+
+# Sample PVC Test
+TEST kubectl create -f $PVC_FILE
+EXPECT_WITHIN -t 60 -v 1 kubectl get pvc | grep $PVC_NAME | grep Bound | wc -l
+
+# Sample app Test
+TEST kubectl create -f $APP_POD_FILE
+EXPECT_WITHIN -t 180 -v 1 kubectl get pods | grep $APP_POD_NAME | grep Running | wc -l
+
+```
+
+## How it works?
+
+Similar to Gluster upstream tests, Binnacle uses [Test Anything
+Protocol(TAP)](https://en.wikipedia.org/wiki/Test_Anything_Protocol). This
+test framework provides syntax sugar on top of the shell script.
+
+Binnacle generates shell script from the given test file and then
+executes the generated script using
+[prove](https://metacpan.org/pod/distribution/Test-Harness/bin/prove)
+command. The following needs to be changed to bash before actually
+running the test file.
+
+## Core functions
+
+Core functions like `TEST`, `EXPECT`, `EXISTS` etc. will be written
+using Python and bash. These functions print the Output in TAP format
+so that prove tool can understand it.
+
+## Binnacle test to Shell script conversion
+
+A few special syntaxes used are not valid bash script. Those need to
+be converted. For example, Pipe character needs to be escaped
+otherwise, `TEST` only runs the first part, and all other commands
+will be run based on the TEST output instead of the first command’s
+Output.
+
+Similarly, the utility file needs to be included in every test file to
+understand `TEST`, `EXPECT`, and other functions.
+
+Loop syntax used in the test file needs to be expanded into a valid
+bash script.
+
+## How is this different from Gluster upstream Tests
+
+We inspired by the Gluster upstream regression tests, which use shell
+scripts for writing test cases. Binnacle provides an additional layer
+on top of shell scripts for ease of use. In addition to ease of use,
+Binnacle also offers the following features compared to the Gluster
+tests framework.
+
+- Multi-node support
+- Plugin support - Extend/add the functionalities easily using Python/Bash
+- Detailed easy to parse Output in verbose mode
+- Ease of use
+
+## Known issues
+
+- The test framework will not work on the Windows operating
+  system. Windows support can be adopted in the future but not a
+  priority for now.
+- Reusing Test cases as-is from other files is not yet supported, but
```

Comment:
Does this mean that a library of tests is not possible at this time?

Reply:
I realized now that, it can be supported as well. I will remove this point.